manifests/service_limits.pp

```diff
@@ -1,51 +1,87 @@
-# -- Define: systemd::service_limits
-# Creates a custom config file and reloads systemd
+# Adds a set of custom limits to the service
+#
+# @api public
+#
+# @see systemd.exec(5)
+#
+# @attr name [Pattern['^.+\.(service|socket|mount|swap)$']]
+#   The name of the service that you will be modifying
+#
+# @param $ensure
+#   Whether to drop a file or remove it
+#
+# @param path
+#   The path to the main systemd settings directory
+#
+# @param limits
+#   A Hash of service limits matching the settings in ``systemd.exec(5)``
+#
+#   * Mutually exclusive with ``$source``
+#
+# @param source
+#   A ``File`` resource compatible ``source``
+#
+#  * Mutually exclusive with ``$limits``
+#
+# @param restart_service
+#   Restart the managed service after setting the limits
+#
 define systemd::service_limits(
-  $ensure          = file,
-  $path            = '/etc/systemd/system',
-  $limits          = undef,
-  $source          = undef,
-  $restart_service = true
+  Enum['file','absent']            $ensure          = 'file',
+  Stdlib::Absolutepath             $path            = '/etc/systemd/system',
+  Optional[Systemd::ServiceLimits] $limits          = undef,
+  Optional[String]                 $source          = undef,
+  Boolean                          $restart_service = true
 ) {
   include ::systemd
 
-  if $limits {
-    validate_hash($limits)
-    $content = template('systemd/limits.erb')
+  if $title !~ Pattern['^.+\.(service|socket|mount|swap)$'] {
+    fail('$name must match Pattern["^.+\.(service|socket|mount|swap)$"]')
+  }
+
+  if $limits and !empty($limits) {
+    $_content = template("${module_name}/limits.erb")
```

Comment:
It denotes it as a private variable that should not be used outside of the class. It's been convention for a while but hasn't made it into the style guide yet.

When actual private variables become a thing, this will be easier.